ocaml/xapi/xapi_pool.ml

```diff
@@ -626,6 +626,138 @@ let create_or_get_pif_on_master __context rpc session_id (pif_ref, pif) : API.re
+        ~mAC ~members:new_members ~primary_slave ~mode ~properties
+    in
+    info "Created new bond=%s on master" (Ref.string_of bond_ref);
+    vlan_pif_device := device;
+    remote_tagged_pif_network := primary_slave_network
+  end
+  else begin
+    (* Get the remote tagged pif network for management on VLAN on physical PIF *)
+    vlan_pif_device := pif.API.pIF_device;
+    let tagged_pif = Db.VLAN.get_tagged_PIF ~__context ~self:vlan_ref in
+    let tagged_pif_network = Db.PIF.get_network ~__context ~self:tagged_pif in
+    let tagged_pif_network_record = Db.Network.get_record ~__context ~self:tagged_pif_network in
+    remote_tagged_pif_network := create_or_get_network_on_master __context rpc session_id
+      (tagged_pif_network, tagged_pif_network_record)
+  end;
+  (* Introduce VLAN PIF on the Pool *)
+  let new_vlan_pif_ref =
+    try Client.PIF.get_by_uuid ~rpc ~session_id ~uuid:pif.API.pIF_uuid
+    with _ ->
+      debug "Found no PIF with uuid = '%s' on the master, so creating one." pif.API.pIF_uuid;
+      Client.PIF.pool_introduce ~rpc ~session_id
+        ~device:!vlan_pif_device
+        ~network:remote_mgmt_network
+        ~host:new_host_ref
+        ~mAC:pif.API.pIF_MAC
+        ~mTU:pif.API.pIF_MTU
+        ~vLAN:pif.API.pIF_VLAN
+        ~physical:pif.API.pIF_physical
+        ~ip_configuration_mode:pif.API.pIF_ip_configuration_mode
+        ~iP:pif.API.pIF_IP
+        ~netmask:pif.API.pIF_netmask
+        ~gateway:pif.API.pIF_gateway
+        ~dNS:pif.API.pIF_DNS
+        ~bond_slave_of:pif.API.pIF_bond_slave_of
+        ~vLAN_master_of:pif.API.pIF_VLAN_master_of
+        ~management:pif.API.pIF_management
+        ~other_config:pif.API.pIF_other_config
+        ~disallow_unplug:pif.API.pIF_disallow_unplug
+        ~ipv6_configuration_mode:pif.API.pIF_ipv6_configuration_mode
+        ~iPv6:pif.API.pIF_IPv6
+        ~ipv6_gateway:pif.API.pIF_ipv6_gateway
+        ~primary_address_type:pif.API.pIF_primary_address_type
+        ~managed:pif.API.pIF_managed
+        ~properties:pif.API.pIF_properties
+  in
+  (* Get the remote tagged PIF - For bonded pif its on the primary slave pif network *)
+  let expr =
+    Printf.sprintf "field \"network\"=\"%s\" and field \"host\"=\"%s\""
+    (Ref.string_of !remote_tagged_pif_network)
+    (Ref.string_of new_host_ref)
+  in
+  let remote_physical_pif =
+    match Client.PIF.get_all_records_where ~rpc ~session_id ~expr with
+    | [] -> Ref.null
+    | (pif, _) :: _ -> pif
+  in
+  let vlan_uuid = Db.VLAN.get_uuid ~__context ~self:vlan_ref in
+  let _ =
+    try Client.VLAN.get_by_uuid ~rpc ~session_id ~uuid:vlan_uuid
+    with _ ->
+      debug "Found no VLAN with uuid = '%s' on the master, so creating one." vlan_uuid;
+      Client.VLAN.pool_introduce ~rpc ~session_id
+        ~tagged_PIF:remote_physical_pif
+        ~untagged_PIF:new_vlan_pif_ref
+        ~tag:vlan_record.API.vLAN_tag
+        ~other_config:vlan_record.API.vLAN_other_config
+  in
+  new_vlan_pif_ref
```

Comment:
This is a very long and complicated function, and I hope we can simplify it a bit. The idea of the `create_or_get_*_on_master` functions, as the comment above `update_non_vm_metadata` implies, is that:

1. they can be called in any order for all DB records that need to be recreated on the pool, and
2. if one depends on the creation of another object, it can just call the `create_or_get` function of that object (e.g. `create_or_get_pif_on_master` needs a network, so it calls `create_or_get_network_on_master`).

It would be nice to keep this structure, just add `create_or_get_vlan_on_master` and `create_or_get_bond_on_master` if required, and call them on the things we want from `update_non_vm_metadata`.

At the moment, the "update PIFs" block calls `create_or_get_pif_on_master` for the physical PIFs only. We can unconditionally include the management PIF in this list (use `Listext.List.setify` to remove any dups), and use it as the starting point. Then we can extend `create_or_get_pif_on_master` to check if it is a VLAN master, and call `create_or_get_vlan_on_master` if so, etc.